src/cargo/ops/cargo_generate_lockfile.rs

```diff
@@ -152,10 +153,17 @@ pub fn update_lockfile(ws: &Workspace<'_>, opts: &UpdateOptions<'_>) -> CargoRes
                 format!("{} -> v{}", removed[0], added[0].version())
             };
 
-            if removed[0].version() > added[0].version() {
-                print_change("Downgrading", msg, &style::WARN)?;
-            } else {
-                print_change("Updating", msg, &style::GOOD)?;
+            let removed = removed[0].version();
+            let added = added[0].version();
+            match Ord::cmp(
+                &(removed.major, removed.minor, removed.patch, &removed.pre),
+                &(added.major, added.minor, added.patch, &added.pre),
+            ) {
+                // Equal: versions differ only in build metadata, which is
+                // considered neither an upgrade nor downgrade.
+                Ordering::Equal => print_change("Switching", msg, &style::NOTE)?,
+                Ordering::Less => print_change("Updating", msg, &style::GOOD)?,
```

Comment:
Maybe calling any build metadata change an "Update" is okay and we don't need a new verb.

```suggestion
                Ordering::Less | Ordering::Equal => print_change("Updating", msg, &style::GOOD)?,
```

Reply:
I think I'd be fine with that (and us simplifying this back down to using the built-in `Ord`).

If there is a strong enough reason to not keep the built-in `Ord`, we should probably consider centralizing this as we have these kind of checks in a couple places and it'd help in making the intent clearer and less risks of bugs imo

Reply:
(oh right, because thats the crux of the problem)

I still lean towards adding a function in `semver_ext.rs` to encapsulate this, updating the other places.

Reply:
I added it directly to `semver` instead.

Still working on auditing the other places that Cargo calls Ord/PartialOrd on Version. To do this, I built cargo with a patch of semver which deletes those impls. Here is the output:

<details>
<summary>`cargo check` output with patched semver</summary>

```console
error[E0369]: binary operation `>` cannot be applied to type `&semver::Version`
   --> src/cargo/core/compiler/future_incompat.rs:352:40
    |
352 |             .filter(|version| *version > pkg_id.version())
    |                               -------- ^ ---------------- &semver::Version
    |                               |
    |                               &semver::Version

error[E0277]: the trait bound `semver::Version: std::cmp::Ord` is not satisfied
   --> src/cargo/core/compiler/future_incompat.rs:354:26
    |
354 |         updated_versions.sort();
    |                          ^^^^ the trait `std::cmp::Ord` is not implemented for `semver::Version`
    |
    = note: required for `&semver::Version` to implement `std::cmp::Ord`
note: required by a bound in `std::slice::<impl [T]>::sort`
   --> /home/david/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/slice.rs:209:12
    |
207 |     pub fn sort(&mut self)
    |            ---- required by a bound in this associated function
208 |     where
209 |         T: Ord,
    |            ^^^ required by this bound in `std::slice::<impl [T]>::sort`

error[E0369]: binary operation `>=` cannot be applied to type `&semver::Version`
    --> src/cargo/core/compiler/job_queue/mod.rs:1148:35
     |
1148 |             || unit.pkg.version() >= &Version::new(1, 4, 8)
     |                ------------------ ^^ ---------------------- &semver::Version
     |                |
     |                &semver::Version

error[E0277]: can't compare `semver::Version` with `semver::Version`
  --> src/cargo/core/package_id.rs:28:5
   |
25 | #[derive(PartialOrd, Eq, Ord)]
   |          ---------- in this derive macro expansion
...
28 |     version: semver::Version,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ no implementation for `semver::Version < semver::Version` and `semver::Version > semver::Version`
   |
   = help: the trait `PartialOrd` is not implemented for `semver::Version`
   = note: this error originates in the derive macro `PartialOrd` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `semver::Version: std::cmp::Ord` is not satisfied
  --> src/cargo/core/package_id.rs:28:5
   |
25 | #[derive(PartialOrd, Eq, Ord)]
   |                          --- in this derive macro expansion
...
28 |     version: semver::Version,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::cmp::Ord` is not implemented for `semver::Version`
   |
   = note: this error originates in the derive macro `Ord` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `semver::Version: std::cmp::Ord` is not satisfied
   --> src/cargo/core/registry.rs:862:14
    |
862 |         vers.sort();
    |              ^^^^ the trait `std::cmp::Ord` is not implemented for `semver::Version`
    |
    = note: required for `&semver::Version` to implement `std::cmp::Ord`
note: required by a bound in `std::slice::<impl [T]>::sort`
   --> /home/david/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/slice.rs:209:12
    |
207 |     pub fn sort(&mut self)
    |            ---- required by a bound in this associated function
208 |     where
209 |         T: Ord,
    |            ^^^ required by this bound in `std::slice::<impl [T]>::sort`

error[E0277]: the trait bound `semver::Version: std::cmp::Ord` is not satisfied
   --> src/cargo/core/registry.rs:916:18
    |
916 |             vers.sort();
    |                  ^^^^ the trait `std::cmp::Ord` is not implemented for `semver::Version`
    |
    = note: required for `&semver::Version` to implement `std::cmp::Ord`
note: required by a bound in `std::slice::<impl [T]>::sort`
   --> /home/david/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/slice.rs:209:12
    |
207 |     pub fn sort(&mut self)
    |            ---- required by a bound in this associated function
208 |     where
209 |         T: Ord,
    |            ^^^ required by this bound in `std::slice::<impl [T]>::sort`

error[E0599]: the method `cmp` exists for reference `&Version`, but its trait bounds were not satisfied
   --> src/cargo/core/resolver/errors.rs:241:52
    |
241 |     candidates.sort_unstable_by(|a, b| b.version().cmp(a.version()));
    |                                                    ^^^ method cannot be called on `&Version` due to unsatisfied trait bounds
    |
   ::: /git/semver/src/lib.rs:163:1
    |
163 | pub struct Version {
    | ------------------
    | |
    | doesn't satisfy `semver::Version: Iterator`
    | doesn't satisfy `semver::Version: std::cmp::Ord`
    |
    = note: the following trait bounds were not satisfied:
            `semver::Version: std::cmp::Ord`
            which is required by `&semver::Version: std::cmp::Ord`
            `&semver::Version: Iterator`
            which is required by `&mut &semver::Version: Iterator`
            `semver::Version: Iterator`
            which is required by `&mut semver::Version: Iterator`

error[E0599]: the method `cmp` exists for reference `&Version`, but its trait bounds were not satisfied
   --> src/cargo/core/resolver/version_prefs.rs:65:43
    |
65  |                     let cmp = a.version().cmp(b.version());
    |                                           ^^^ method cannot be called on `&Version` due to unsatisfied trait bounds
    |
   ::: /git/semver/src/lib.rs:163:1
    |
163 | pub struct Version {
    | ------------------
    | |
    | doesn't satisfy `semver::Version: Iterator`
    | doesn't satisfy `semver::Version: std::cmp::Ord`
    |
    = note: the following trait bounds were not satisfied:
            `semver::Version: std::cmp::Ord`
            which is required by `&semver::Version: std::cmp::Ord`
            `&semver::Version: Iterator`
            which is required by `&mut &semver::Version: Iterator`
            `semver::Version: Iterator`
            which is required by `&mut semver::Version: Iterator`

error[E0277]: the trait bound `semver::Version: std::cmp::Ord` is not satisfied
   --> src/cargo/ops/cargo_add/mod.rs:556:27
    |
556 |             possibilities.sort_by_key(|s| {
    |                           ^^^^^^^^^^^ the trait `std::cmp::Ord` is not implemented for `semver::Version`
    |
    = help: the following other types implement trait `std::cmp::Ord`:
              ()
              (C, B, A, Z, Y, X, W, V, U, T)
              (T,)
              (D, C, B, A, Z, Y, X, W, V, U, T)
              (E, D, C, B, A, Z, Y, X, W, V, U, T)
              (U, T)
              (B, A, Z, Y, X, W, V, U, T)
              (W, V, U, T)
            and 5 others
    = note: required for `(bool, semver::Version)` to implement `std::cmp::Ord`
note: required by a bound in `std::slice::<impl [T]>::sort_by_key`
   --> /home/david/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/slice.rs:308:12
    |
305 |     pub fn sort_by_key<K, F>(&mut self, mut f: F)
    |            ----------- required by a bound in this associated function
...
308 |         K: Ord,
    |            ^^^ required by this bound in `std::slice::<impl [T]>::sort_by_key`

error[E0277]: the trait bound `semver::Version: std::cmp::Ord` is not satisfied
    --> src/cargo/ops/cargo_add/mod.rs:596:34
     |
596  | ...                   .max_by_key(|s| s.version());
     |                        ^^^^^^^^^^ the trait `std::cmp::Ord` is not implemented for `semver::Version`
     |
     = note: required for `&semver::Version` to implement `std::cmp::Ord`
note: required by a bound in `std::iter::Iterator::max_by_key`
    --> /home/david/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:3262:22
     |
3262 |     fn max_by_key<B: Ord, F>(self, f: F) -> Option<Self::Item>
     |                      ^^^ required by this bound in `Iterator::max_by_key`

error[E0369]: binary operation `<` cannot be applied to type `&semver::Version`
   --> src/cargo/ops/cargo_add/mod.rs:604:46
    |
604 |                     if latest_msrv.version() < latest.version() {
    |                        --------------------- ^ ---------------- &semver::Version
    |                        |
    |                        &semver::Version

error[E0277]: the trait bound `semver::Version: std::cmp::Ord` is not satisfied
    --> src/cargo/ops/cargo_add/mod.rs:891:10
     |
891  |         .min_by_key(|s| {
     |          ^^^^^^^^^^ the trait `std::cmp::Ord` is not implemented for `semver::Version`
     |
     = help: the following other types implement trait `std::cmp::Ord`:
               ()
               (C, B, A, Z, Y, X, W, V, U, T)
               (T,)
               (D, C, B, A, Z, Y, X, W, V, U, T)
               (E, D, C, B, A, Z, Y, X, W, V, U, T)
               (U, T)
               (B, A, Z, Y, X, W, V, U, T)
               (W, V, U, T)
             and 5 others
     = note: required for `&semver::Version` to implement `std::cmp::Ord`
     = note: 1 redundant requirement hidden
     = note: required for `(bool, &semver::Version)` to implement `std::cmp::Ord`
note: required by a bound in `std::iter::Iterator::min_by_key`
    --> /home/david/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:3324:22
     |
3324 |     fn min_by_key<B: Ord, F>(self, f: F) -> Option<Self::Item>
     |                      ^^^ required by this bound in `Iterator::min_by_key`

error[E0277]: the trait bound `semver::Version: std::cmp::Ord` is not satisfied
    --> src/cargo/ops/cargo_add/mod.rs:1082:10
     |
1082 |         .max()
     |          ^^^ the trait `std::cmp::Ord` is not implemented for `semver::Version`
     |
note: the method call chain might not have had the expected associated types
    --> src/cargo/ops/cargo_add/mod.rs:1070:10
     |
1055 | /     version_req
1056 | |         .comparators
     | |____________________- this expression has type `Vec<Comparator>`
1057 |           .iter()
     |            ------ `Iterator::Item` is `&Comparator` here
1058 |           .filter(|c| {
     |  __________-
1059 | |             matches!(
1060 | |                 c.op,
1061 | |                 // Only ops we can determine a precise version from
...    |
1068 | |             )
1069 | |         })
     | |__________- `Iterator::Item` remains `&Comparator` here
1070 |           .filter_map(|c| {
     |  __________^
1071 | |             // Only do it when full precision is specified
1072 | |             c.minor.and_then(|minor| {
1073 | |                 c.patch.map(|patch| semver::Version {
...    |
1080 | |             })
1081 | |         })
     | |__________^ `Iterator::Item` changed to `Version` here
note: required by a bound in `std::iter::Iterator::max`
    --> /home/david/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:3203:21
     |
3200 |     fn max(self) -> Option<Self::Item>
     |        --- required by a bound in this associated function
...
3203 |         Self::Item: Ord,
     |                     ^^^ required by this bound in `Iterator::max`

error[E0599]: the method `partial_cmp` exists for reference `&Version`, but its trait bounds were not satisfied
   --> src/cargo/ops/cargo_compile/mod.rs:802:54
    |
802 |                 units.sort_by(|a, b| a.pkg.version().partial_cmp(b.pkg.version()).unwrap());
    |                                                      ^^^^^^^^^^^ method cannot be called on `&Version` due to unsatisfied trait bounds
    |
   ::: /git/semver/src/lib.rs:163:1
    |
163 | pub struct Version {
    | ------------------
    | |
    | doesn't satisfy `semver::Version: Iterator`
    | doesn't satisfy `semver::Version: PartialOrd<_>`
    |
    = note: the following trait bounds were not satisfied:
            `semver::Version: PartialOrd<_>`
            which is required by `&semver::Version: PartialOrd<&_>`
            `&semver::Version: Iterator`
            which is required by `&mut &semver::Version: Iterator`
            `semver::Version: Iterator`
            which is required by `&mut semver::Version: Iterator`

error[E0369]: binary operation `<` cannot be applied to type `&semver::Version`
   --> src/cargo/ops/cargo_compile/mod.rs:806:40
    |
806 |                     unit.pkg.version() < &newest_version
    |                     ------------------ ^ --------------- &semver::Version
    |                     |
    |                     &semver::Version

error[E0369]: binary operation `<` cannot be applied to type `&semver::Version`
   --> src/cargo/ops/fix.rs:331:49
    |
331 |         pid.name() == "diesel" && pid.version() < &Version::new(1, 4, 8)
    |                                   ------------- ^ ---------------------- &semver::Version
    |                                   |
    |                                   &semver::Version

error[E0277]: the trait bound `semver::Version: std::cmp::Ord` is not satisfied
   --> src/cargo/ops/vendor.rs:177:38
    |
177 |         if let Some(prev) = map.get(&id.version()) {
    |                                 ---  ^^^^^^^^^^^^ the trait `std::cmp::Ord` is not implemented for `semver::Version`
    |                                 |
    |                                 required by a bound introduced by this call
    |
    = note: required for `&semver::Version` to implement `std::cmp::Ord`
note: required by a bound in `BTreeMap::<K, V, A>::get`
   --> /home/david/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/collections/btree/map.rs:697:12
    |
694 |     pub fn get<Q: ?Sized>(&self, key: &Q) -> Option<&V>
    |            --- required by a bound in this associated function
...
697 |         Q: Ord,
    |            ^^^ required by this bound in `BTreeMap::<K, V, A>::get`

error[E0277]: the trait bound `semver::Version: std::cmp::Ord` is not satisfied
   --> src/cargo/ops/vendor.rs:177:33
    |
177 |         if let Some(prev) = map.get(&id.version()) {
    |                                 ^^^ the trait `std::cmp::Ord` is not implemented for `semver::Version`
    |
    = note: required for `&semver::Version` to implement `std::cmp::Ord`
note: required by a bound in `BTreeMap::<K, V, A>::get`
   --> /home/david/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/collections/btree/map.rs:696:24
    |
694 |     pub fn get<Q: ?Sized>(&self, key: &Q) -> Option<&V>
    |            --- required by a bound in this associated function
695 |     where
696 |         K: Borrow<Q> + Ord,
    |                        ^^^ required by this bound in `BTreeMap::<K, V, A>::get`

error: could not compile `cargo` (lib) due to 19 previous errors
```
</details>

Reply:
https://github.com/rust-lang/cargo/commit/892890c3446f4efd6782e4adb766263ee52aac6e: I found these 2 places that might want to ignore build metadata when ordering versions, but I was not clear on the intended behavior or whether it matters.

Reply:
The cases I was thinking of were
- https://github.com/rust-lang/cargo/blob/master/src/cargo/sources/registry/index.rs#L677
- https://github.com/rust-lang/cargo/blob/master/src/cargo/util/semver_ext.rs#L86

An odd case is https://github.com/rust-lang/cargo/blob/master/src/cargo/core/package_id.rs#L47 where it doesn't use `==` because it claims its trying to include `build` which implies the outer type should be ignoring `build` but it uses `==`.

Reply:
I added a test (`unequal_build_metadata` in package_id.rs) so that if Cargo were to adopt a different semver library that defines `==` for Version differently, that PackageIdInner comparison does not accidentally change behavior.

I agree with your impression that the PackageId comparison is not likely to be doing what was intended.